### PR TITLE
Image Feature Extraction docs

### PR DESCRIPTION
**Pull request by merveenoyan**
This PR adds task guide for image feature extraction.

Note that as of now the image feature extraction pipeline doesn't have a pooler output, so the result in the doc is the theoretical result :') This PR can be reviewed after that change is merged since this wouldn't work without `ViTPooler`.

### CODE REVIEW

**Comment by amyeroberts** [comment]
@merveenoyan There's a PR to add the pool option - I'm just waiting for review #28985

File: docs/source/en/tasks/image_feature_extraction.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# Image Feature Extraction
+
+[[open-in-colab]]
+
+Image feature extraction is the task of extracting semantically meaningful features given an image. This has many use cases, including, image similarity and image retrieval. Moreover, every computer vision model can be used for image feature extraction, where one can remove the task-specific head (image classification, object detection etc) and get the features. These features are very useful on a higher level: edge detection, corner detection and so on, as well as containing information about the real world (e.g. how a cat looks like) depending on how deep the model is. Therefore, these outputs can be used to train new classifiers on a specific dataset.
```
Comment (amyeroberts) [inline docs/source/en/tasks/image_feature_extraction.md:21]:
```suggestion
Image feature extraction is the task of extracting semantically meaningful features given an image. This has many use cases, including, image similarity and image retrieval. Moreover, every computer vision model can be used for image feature extraction, where one can remove the task-specific head (image classification, object detection etc) and get the features. These features are very useful on a higher level: edge detection, corner detection and so on, as well as containing information about the real world (e.g. what a cat looks like) depending on how deep the model is. Therefore, these outputs can be used to train new classifiers on a specific dataset.
```

**Review by amyeroberts — COMMENTED** [review]
Looks great! 

#28985 has now been merged in, so you can now use the pool option

File: docs/source/en/tasks/image_feature_extraction.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+We can also use `AutoModel` class of transformers to get the features. `AutoModel` loads any transformers model with no task-specific head, and we can use this to get the features.
+
+```python
+from transformers import AutoImageProcessor, AutoModel
+
+processor = AutoImageProcessor.from_pretrained("google/vit-base-patch16-224")
+model = AutoModel.from_pretrained("google/vit-base-patch16-224").to(DEVICE)
+```
+
+Let's write a simple function for inference. We will pass the inputs to the `processor` first and pass it's outputs to the `model`.
```
Comment (amyeroberts) [inline docs/source/en/tasks/image_feature_extraction.md:81]: nit

```suggestion
Let's write a simple function for inference. We will pass the inputs to the `processor` first and pass its outputs to the `model`.
```

File: docs/source/en/tasks/image_feature_extraction.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+pipe = pipeline(task="image-feature-extraction", model_name="google/vit-base-patch16-384", device=DEVICE, pool=True)
+```
+
+We can now infer with `pipe`, by passing in both of the images.
+
+```python
+outputs = pipe([image_real, image_gen])
+```
+
+The output is contains embeddings of those two images. To get the similarity, we need to pass them to a similarity function.
```
Comment (amyeroberts) [inline docs/source/en/tasks/image_feature_extraction.md:57]:
```suggestion
The output contains embeddings of those two images. To get the similarity, we need to pass them to a similarity function. 
```

File: docs/source/en/tasks/image_feature_extraction.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# Image Feature Extraction
+
+[[open-in-colab]]
+
+Image feature extraction is the task of extracting semantically meaningful features given an image. This has many use cases, including, image similarity and image retrieval. Moreover, every computer vision model can be used for image feature extraction, where one can remove the task-specific head (image classification, object detection etc) and get the features. These features are very useful on a higher level: edge detection, corner detection and so on, as well as containing information about the real world (e.g. what a cat looks like) depending on how deep the model is. Therefore, these outputs can be used to train new classifiers on a specific dataset.
```
Comment (amyeroberts) [inline docs/source/en/tasks/image_feature_extraction.md:21]: Nit - I still don't think we should say every CV model can be used for this, because they don't all have a task-specific head that can be removed like this

File: docs/source/en/tasks/image_feature_extraction.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+```python
+from PIL import Image
+import requests
+
+img_urls = ["https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cats.png", "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cats.jpeg"]
+image_real = Image.open(requests.get(img_urls[0], stream=True).raw).convert("RGB")
+image_gen = Image.open(requests.get(img_urls[1], stream=True).raw).convert("RGB")
+```
+
+Let's see the pipeline in action. We will initialize the pipeline. If we do not pass any model, it will be automatically initialized with [google/vit-base-patch16-224](google/vit-base-patch16-224). If you'd like to calculate similarity, set `pool` to True.
```
Comment (amyeroberts) [inline docs/source/en/tasks/image_feature_extraction.md:41]: It would be good to explain a little bit here why you need `pool` to do similarity here, and what is output if `pool=False`

**Review by amyeroberts — COMMENTED** [review]
Looking good! 

Just some small comments

File: docs/source/en/tasks/image_feature_extraction.md (hunk 4 above)
Comment (merveenoyan) [reply docs/source/en/tasks/image_feature_extraction.md:21]: sorry I forgot to address this in previous round

File: docs/source/en/tasks/image_feature_extraction.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+```python
+import torch
+from transformers import pipeline
+
+DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+pipe = pipeline(task="image-feature-extraction", model_name="google/vit-base-patch16-384", device=DEVICE, pool=True)
+```
+
+We can now infer with `pipe`, by passing in both of the images.
```
Comment (MKhalusova) [inline docs/source/en/tasks/image_feature_extraction.md:51]:
```suggestion
To infer with `pipe` pass both images to it.
```

File: docs/source/en/tasks/image_feature_extraction.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+```python
+outputs = pipe([image_real, image_gen])
+```
+
+The output contains pooled embeddings of those two images.
+
+```python
+# get the length of a single output
+print(len(outputs[0][0]))
+# show outputs
```
Comment (MKhalusova) [inline docs/source/en/tasks/image_feature_extraction.md:62]: I don't think this comment is necessary.

File: docs/source/en/tasks/image_feature_extraction.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+print(outputs)
+
+# 768
+# [[[-0.03909236937761307, 0.43381670117378235, -0.06913255900144577,
+```
+
+To get the similarity, we need to pass them to a similarity function.
```
Comment (MKhalusova) [inline docs/source/en/tasks/image_feature_extraction.md:69]:
```suggestion
To get the similarity score, we need to pass them to a similarity function. 
```

File: docs/source/en/tasks/image_feature_extraction.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+similarity_score = cosine_similarity(torch.Tensor(outputs[0]),
+                                     torch.Tensor(outputs[1]), dim=1)
+
+print(similarity_score)
+
+# tensor([0.6043])
+```
+
+If you want to get the last hidden states before pooling, simply avoid passing any value for the `pool` parameter as it is set to `False` by default.
```
Comment (MKhalusova) [inline docs/source/en/tasks/image_feature_extraction.md:82]: Why would I want to get the last hidden states?

File: docs/source/en/tasks/image_feature_extraction.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+```python
+pipe = pipeline(task="image-feature-extraction", model_name="google/vit-base-patch16-224", device=DEVICE)
+output = pipe(image_real)
+```
+Since the outputs are unpooled, we get last hidden states where first dimension is the batch size and the last two are embedding shape.
```
Comment (MKhalusova) [inline docs/source/en/tasks/image_feature_extraction.md:88]:
```suggestion

Since the outputs are unpooled, we get last hidden states where the first dimension is the batch size and the last two are the embedding shape.
```

**Review by MKhalusova — COMMENTED** [review]
Great guide, @merveenoyan ! I've added a few suggestions.

File: docs/source/en/tasks/image_feature_extraction.md (hunk 7 above)
Comment (merveenoyan) [reply docs/source/en/tasks/image_feature_extraction.md:62]: I think it's useful since people can see the pipeline returns outputs for both inputs in a list and the inner list is also a list (often times it's numpy array)

File: docs/source/en/tasks/image_feature_extraction.md (hunk 9 above)
Comment (merveenoyan) [reply docs/source/en/tasks/image_feature_extraction.md]: Addressed this

**Review by MKhalusova — APPROVED** [review]
LGTM, thank you @merveenoyan !

File: docs/source/en/tasks/image_feature_extraction.md [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+# Image Feature Extraction
+
+[[open-in-colab]]
+
+Image feature extraction is the task of extracting semantically meaningful features given an image. This has many use cases, including image similarity and image retrieval. Moreover, most computer vision models can be used for image feature extraction, where one can remove the task-specific head (image classification, object detection etc) and get the features. These features are very useful on a higher level: edge detection, corner detection and so on. They may also contain information about the real world (e.g. what a cat looks like) depending on how deep the model is. Therefore, these outputs can be used to train new classifiers on a specific dataset.
+
+In this guide, you will:
+
+-  Learn to build a simple image similarity system on top of the `image-feature-extraction` pipeline.
```
Comment (amyeroberts) [inline docs/source/en/tasks/image_feature_extraction.md:25]: nit

```suggestion
- Learn to build a simple image similarity system on top of the `image-feature-extraction` pipeline.
```

**Review by amyeroberts — APPROVED** [review]
Looks great - thanks for adding! 🥳
